### PR TITLE
api added for downloader utility data.

### PR DESCRIPTION
**Pull request by Raheela1024**

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -23,6 +24,26 @@ class StatusBody(BaseModel):
...
+            rank = session.write_transaction(self._get_rank, param)
+            return rank
+
+    @staticmethod
+    def _get_rank(tx, param):
+        result = tx.run(
+            'MATCH (parent:Taxonomies)-[:CHILD]->(child:Taxonomies) where parent.name=~' '"' '.*' + param + '.*' '"'
+                                                                                                            'RETURN parent')
+        return result.single()[0]
+
```
Comment (Alexey-ebi) [inline app/main.py]: can you move it to the separate file please

File: app/main.py [hunk 2]
```diff
@@ -23,6 +24,26 @@ class StatusBody(BaseModel):
...
+    def get_rank(self, param):
+        with self.driver.session() as session:
+            rank = session.write_transaction(self._get_rank, param)
+            return rank
+
+    @staticmethod
+    def _get_rank(tx, param):
+        result = tx.run(
+            'MATCH (parent:Taxonomies)-[:CHILD]->(child:Taxonomies) where parent.name=~' '"' '.*' + param + '.*' '"'
```
Comment (Alexey-ebi) [inline app/main.py]: this string is too long, for string interpolation please use this: f"some string {param} some string"

File: app/main.py [hunk 3]
```diff
@@ -124,3 +145,51 @@ async def login(form_data: OAuth2PasswordRequestForm = Depends()):
         "access_token": create_access_token(user['username']),
         "refresh_token": create_refresh_token(user['username']),
     }
+
+
+
+@app.get("/downloader_utility_data")
```
Comment (Alexey-ebi) [inline app/main.py]: too many blank lines

File: app/main.py [hunk 4]
```diff
@@ -124,3 +145,51 @@ async def login(form_data: OAuth2PasswordRequestForm = Depends()):
...
+async def index(taxonomyFilter: str, data_status: str, experiment_type: str):
+    neofourJ = NeoFourJ("bolt://45.88.80.141:30087", "neo4j", "DtolNeo4jAdminUser@123")
```
Comment (Alexey-ebi) [inline app/main.py]: variable in function should be lowercase

File: app/main.py [hunk 5]
```diff
@@ -124,3 +145,51 @@ async def login(form_data: OAuth2PasswordRequestForm = Depends()):
...
+    if experiment_type != '' and experiment_type is not None:
+        query_param = query_param + ',{ "nested" : { "path": "experiment", "query" : { "bool" : { "must" : [' \
+                                        '{ "term" : { "experiment.library_construction_protocol.keyword" : ' + \
+                          '"' + experiment_type + '"' '}}]}}}}'
+
+    query_param = query_param + '] }}}'
+
+    response = es.search(index="data_portal", size=10000, body=query_param)
+    neofourJ.close()
+    return response
```
Comment (Alexey-ebi) [inline app/main.py]: no new line at the end of file

File: app/main.py [hunk 6]
```diff
@@ -124,3 +145,51 @@ async def login(form_data: OAuth2PasswordRequestForm = Depends()):
...
+        query_param = query_param + ',{ "nested": {"path": "genome_notes","query": {"bool": {"must": [{"exists": ' \
+                                            '{"field": "genome_notes.url"}}]}}}} '
+    if experiment_type != '' and experiment_type is not None:
+        query_param = query_param + ',{ "nested" : { "path": "experiment", "query" : { "bool" : { "must" : [' \
+                                        '{ "term" : { "experiment.library_construction_protocol.keyword" : ' + \
+                          '"' + experiment_type + '"' '}}]}}}}'
+
+    query_param = query_param + '] }}}'
+
+    response = es.search(index="data_portal", size=10000, body=query_param)
```
Comment (Alexey-ebi) [inline app/main.py]: this won't work in the near future - we need to have a method that will parse all available data with 10000 sliding windows

File: app/main.py [hunk 7]
```diff
@@ -124,3 +145,51 @@ async def login(form_data: OAuth2PasswordRequestForm = Depends()):
...
+        1].strip() + '"'']}}'
+    elif split_array and split_array[0].strip() == 'Genome Notes':
+        query_param = query_param + ',{ "nested": {"path": "genome_notes","query": {"bool": {"must": [{"exists": ' \
+                                            '{"field": "genome_notes.url"}}]}}}} '
+    if experiment_type != '' and experiment_type is not None:
+        query_param = query_param + ',{ "nested" : { "path": "experiment", "query" : { "bool" : { "must" : [' \
+                                        '{ "term" : { "experiment.library_construction_protocol.keyword" : ' + \
+                          '"' + experiment_type + '"' '}}]}}}}'
+
+    query_param = query_param + '] }}}'
```
Comment (Alexey-ebi) [inline app/main.py]: 1. lots of problems with indentation 
2. will be better to use string interpolation

File: app/main.py [hunk 8]
```diff
@@ -124,3 +145,51 @@ async def login(form_data: OAuth2PasswordRequestForm = Depends()):
         "access_token": create_access_token(user['username']),
         "refresh_token": create_refresh_token(user['username']),
     }
+
+
+
+@app.get("/downloader_utility_data")
+async def index(taxonomyFilter: str, data_status: str, experiment_type: str):
```
Comment (Alexey-ebi) [inline app/main.py]: argument name should be lowercase

**Review by Alexey-ebi — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 1 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py (hunk 3 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py (hunk 4 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py (hunk 8 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py (hunk 7 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py (hunk 5 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py (hunk 2 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py (hunk 6 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py [hunk 9]
```diff
@@ -124,3 +125,101 @@ async def login(form_data: OAuth2PasswordRequestForm = Depends()):
         "access_token": create_access_token(user['username']),
         "refresh_token": create_refresh_token(user['username']),
     }
+
+
+@app.get("/downloader_utility_data/")
+async def downloader_utility_data(taxonomy_filter: str, data_status: str, experiment_type: str):
+    global split_array
```
Comment (Alexey-ebi) [inline app/main.py]: why do you need to use global variable here?

File: app/main.py [hunk 10]
```diff
@@ -124,3 +125,101 @@ async def login(form_data: OAuth2PasswordRequestForm = Depends()):
...
+        }
+        body["query"]["bool"]["filter"].append(nested_dict)
+    response = es.search(index="data_portal", from_=0, size=10000, body=body)
+    total_count = response['hits']['total']['value']
+    result = response['hits']['hits']
+    results_count = len(response['hits']['hits'])
+    while total_count > results_count:
+        response1 = es.search(index="data_portal", from_=results_count, size=10000, body=body)
+        result.append(response1['hits']['hits'])
+        results_count += len(response1['hits']['hits'])
```
Comment (Alexey-ebi) [inline app/main.py]: This won't work once you get more than 10k samples

**Review by Alexey-ebi — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 9 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

File: app/main.py (hunk 10 above)
Comment (Raheela1024) [reply app/main.py]: Resolved

**Pull request merged**